flink-formats/flink-csv/src/test/java/org/apache/flink/formats/csv/CsvBulkWriterIT.java

```diff
@@ -0,0 +1,205 @@
+ */
+
+package org.apache.flink.formats.csv;
+
+import org.apache.flink.api.common.eventtime.WatermarkStrategy;
+import org.apache.flink.api.common.restartstrategy.RestartStrategies;
+import org.apache.flink.api.common.serialization.BulkWriter;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.connector.datagen.source.DataGeneratorSource;
+import org.apache.flink.connector.datagen.source.TestDataGenerators;
+import org.apache.flink.connector.file.sink.FileSink;
+import org.apache.flink.core.fs.FSDataOutputStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+
+import org.apache.commons.io.FileUtils;
+import org.jetbrains.annotations.NotNull;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class CsvBulkWriterIT {
+
+    @TempDir File outDir;
+
+    @Test
+    public void testNoDataIsWrittenBeforeFlush() throws Exception {
+
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setParallelism(1);
+        env.enableCheckpointing(100);
+        env.setRestartStrategy(RestartStrategies.noRestart());
+
+        // Workaround serialization limitations
+        File outDirRef = new File(outDir.getAbsolutePath());
+
+        FileSink<Pojo> sink =
+                FileSink.forBulkFormat(
+                                new org.apache.flink.core.fs.Path(outDir.getAbsolutePath()),
+                                out -> {
+                                    FSDataOutputStreamWrapper outputStreamWrapper =
+                                            new FSDataOutputStreamWrapper(out);
+                                    return new CsvBulkWriterWrapper<>(
+                                            CsvBulkWriter.forPojo(Pojo.class, outputStreamWrapper),
+                                            outputStreamWrapper,
+                                            outDirRef);
+                                })
+                        .build();
+
+        List<Pojo> integers = Arrays.asList(new Pojo(1), new Pojo(2));
+        DataGeneratorSource<Pojo> generatorSource =
+                TestDataGenerators.fromDataWithSnapshotsLatch(
+                        integers, TypeInformation.of(Pojo.class));
+        env.fromSource(generatorSource, WatermarkStrategy.noWatermarks(), "").sinkTo(sink);
+        env.execute();
+        assertThat(getResultsFromSinkFiles(outDir)).containsSequence("1", "2", "1", "2");
+    }
+
+    private static class CsvBulkWriterWrapper<T> implements BulkWriter<T> {
+
+        private static int callCounter = 0;
+
+        private static int unFlushedCounter = 0;
```

Comment:
nitpick: I wonder if it would make the logic slightly easier to read if you change this from counting the unflushed elements, to tracking the expected number of flushed elements. Eg:

```java
private static int addedElements = 0;
private static int expectedFlushedElements = 0;
        
@Override
public void addElement(T element) throws IOException {
    addedElements++;
    csvBulkWriter.addElement(element);
    
    assertThat(getResultsFromSinkFiles(outDir)).hasSize(expectedFlushedElements);
}

@Override
public void flush() throws IOException {
    csvBulkWriter.flush();
    expectedFlushedElements = addedElements;

    assertThat(getResultsFromSinkFiles(outDir)).hasSize(expectedFlushedElements);
}
```